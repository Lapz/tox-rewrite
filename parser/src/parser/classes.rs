use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_class(&mut self, checkpoint: rowan::Checkpoint) {
        self.start_node_at(checkpoint, CLASS_DEF);

        self.expect(CLASS_KW);

        self.ident();

        if self.at(T![<]) {
            self.parse_type_params(false);
        }

        self.parse_class_body();

        self.finish_node()
    }

    fn parse_class_body(&mut self) {
        self.expect(T!["{"]);

        while !self.at(EOF) && !self.at(T!["}"]) {
            let has_visibility = self.has_visibility();

            if has_visibility {
                let checkpoint = self.checkpoint();
                self.parse_visibility();

                match self.current() {
                    IDENT => self.parse_named_field(),
                    T![fn] => self.parse_function(checkpoint),

                    _ => self.error(
                        "Expected an identifier | `pub` | `fn` ",
                        format!(
                            "Expected an identifier | `pub` | `fn`  but instead found `{}`",
                            self.current_string()
                        ),
                    ),
                }
            } else {
                let checkpoint = self.checkpoint();
                match self.current() {
                    IDENT => self.parse_named_field(),
                    T![fn] => self.parse_function(checkpoint),

                    _ => self.error(
                        "Expected an identifier | `pub` | `fn` ",
                        format!(
                            "Expected an identifier | `pub` | `fn`  but instead found `{}`",
                            self.current_string()
                        ),
                    ),
                }
            }
        }

        self.expect(T!["}"]);
    }

    fn parse_named_field(&mut self) {
        self.start_node(NAMED_FIELD_DEF);
        self.ident();
        self.expect(T![:]);
        self.parse_type();
        self.expect(T![;]);
        self.finish_node();
    }
}

#[cfg(test)]
mod tests {

    use crate::utils::parse;
    use syntax::{ClassDefOwner, NamedFieldsOwner};
    #[test]
    fn test_class_fields() {
        let source_file = parse("class Person { name:String; surname:String;}");

        let class = source_file.classes().next().unwrap();

        assert_eq!(class.fields().count(), 2)
    }
    test_parser! {parse_empty_class,"class Foo {}"}
    test_parser! {parse_exported_class,"export class Foo {}"}
    test_parser! {parse_class_generic,"class Result<T,E> {}"}
    test_parser! {parse_class_fields,"class Person { name:String; surname:String;}"}
    test_parser! {parse_class_fields_methods,"class Person { name:String; surname:String; fn hello(self) {}}"}
    test_parser! {parse_class_methods,"class Person { name:String; surname:String; fn new() -> Person {}}"}
}
