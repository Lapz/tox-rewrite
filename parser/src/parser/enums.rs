use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_enum(&mut self, checkpoint: rowan::Checkpoint) {
        self.start_node_at(checkpoint, ENUM_DEF);

        self.expect(T![enum]);

        self.ident();

        if self.at(T![<]) {
            self.parse_type_params(false);
        }

        self.parse_enum_variants();

        self.finish_node()
    }

    fn parse_enum_variants(&mut self) {
        self.start_node(ENUM_VARIANT_LIST);

        self.expect(T!["{"]);

        while !self.at(EOF) && !self.at(T!["}"]) {
            self.parse_enum_variant();
            if !self.at(T!["}"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T!["}"]);

        self.finish_node();
    }

    fn parse_enum_variant(&mut self) {
        self.start_node(ENUM_VARIANT);

        self.ident();

        if self.at(T!["("]) {
            self.parse_enum_variant_types()
        }

        self.finish_node();
    }

    fn parse_enum_variant_types(&mut self) {
        self.bump(); // Eat the "("
        while !self.at(EOF) && !self.at(T![")"]) {
            self.parse_type();

            if !self.at(T![")"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![")"]);
    }
}

#[cfg(test)]
mod tests {
    test_parser! {parse_empty_enum,"enum Foo {}"}
    test_parser! {parse_exported_enum,"export enum Foo {}"}
    test_parser! {parse_enum_with_variants,"enum Foo {A,B,C}"}
    test_parser! {parse_enum_variants_trailing_comma,"enum Foo {A,B,C,}"}
    test_parser! {parse_enum_variants_types,"enum Foo { A([i32]),B(i32),C((i32,i32)),}"}
    test_parser! {parse_enum_generic,"enum Result<T,E> { Ok(T),Err(E)}"}
}
