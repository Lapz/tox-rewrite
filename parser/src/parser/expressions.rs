use crate::parser::{
    pratt::{Precedence, Rule, RuleToken},
    Parser, Restrictions,
};

use syntax::T;

mod binary;
mod block;
mod break_expr;
mod call_expr;
mod closure_expr;
mod continue_expr;
mod do_expr;

mod field_expr;
mod for_expr;
mod grouping;
mod ident;
mod if_expr;
mod index_expr;
mod let_expr;
mod literal;
mod match_expr;
mod record_expr;
mod return_expr;
mod unary;
mod while_expr;

pub use binary::BinaryParselet;
pub use call_expr::CallParselet;
pub use closure_expr::ClosureParselet;
pub use field_expr::FieldParselet;
pub use grouping::GroupingParselet;
pub use ident::IdentParselet;
pub use index_expr::IndexParselet;
pub use literal::LiteralParselet;
pub use record_expr::RecordParselet;
pub use unary::UnaryParselet;

impl<'a> Parser<'a> {
    pub(crate) fn parse_expression(&mut self, precedence: Precedence, restrictions: Restrictions) {
        let check_point = self.builder.checkpoint();

        let token = self.current();

        let mut rule = token.rule();

        let parser = self.prefix.get(&rule);

        let parser = if let Some(parser) = parser {
            parser
        } else {
            self.error(
                "Expected an expression",
                format!(
                    "Expected an expression but instead found `{}`",
                    self.current_string()
                ),
            );
            return;
        };

        parser.parse(self);

        while precedence <= self.precedence() {
            {
                let token = self.current();
                rule = token.rule();
            }

            let parser = self.infix.get(&rule);

            let parser = if parser.is_some() {
                parser.unwrap()
            } else {
                self.error(
                    "Expected an expression",
                    format!(
                        "Expected an expression but instead found `{}`",
                        self.current_string()
                    ),
                );
                break;
            };

            if restrictions.forbid_record && rule == RuleToken::LBrace {
                break;
            }
            parser.parse(self, check_point);
        }

        if precedence == Precedence::None {
            self.recover_until(T![;]);
        }
    }
}
