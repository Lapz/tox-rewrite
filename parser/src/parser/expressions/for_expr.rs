use syntax::T;

use crate::parser::{Parser, Precedence, Restrictions};
use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_for_expr(&mut self) {
        self.start_node(FOR_EXPR);

        self.expect(T![for]);

        self.expect(T!["("]);

        if self.at(T![;]) {
            self.bump()
        } else if self.at(T![let]) {
            self.parse_let_expr();
            self.expect(T![;]);
        } else {
            self.parse_expression(Precedence::Assignment, Restrictions::default());
            self.expect(T![;]);
        }

        if self.at(T![;]) {
            self.bump()
        } else {
            self.parse_expression(Precedence::Comparison, Restrictions::default());
            self.expect(T![;]);
        }

        if self.at(T![;]) {
            self.bump();
        } else if self.at(T![")"]) {
        } else {
            self.parse_expression(Precedence::Assignment, Restrictions::default());
        }

        self.expect(T![")"]);

        self.parse_block();

        self.finish_node()
    }
}

mod tests {
    test_parser! {parse_empty_for_expr,"fn main() {for(;;;) {}}"}
    test_parser! {parse_for_expr,"
                    fn main() {
                        for(let x = 10; x< 10; x=x+1) {
                            println(\"it works \")
                        };}"
    }
    test_parser! {parse_for_no_incr,"fn main() { for (let x=10;x<10;;){};}"}
    test_parser! {parse_for_no_cond,"fn main() { for (let x=10;;x+=10){};}"}
    test_parser! {parse_for_no_init,"fn main() { for (;x<10;x+=10){};}"}
}
