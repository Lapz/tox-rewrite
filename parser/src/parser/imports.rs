use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_import(&mut self) {
        self.start_node(IMPORT_DEF);

        self.expect(T![import]);

        while !self.at(EOF) && !self.at(T![;]) {
            self.parse_import_segment();

            if !self.at(T![;]) && !self.expected(T![::]) {
                break;
            }
        }

        self.expect(T![;]);

        self.finish_node();
    }

    pub(crate) fn parse_import_segment(&mut self) {
        if self.at(T!["{"]) {
            self.start_node(IMPORT_LIST);

            self.bump();

            while !self.at(EOF) && !self.at(T!["}"]) {
                self.parse_import_segment();

                if !self.at(T!["}"]) && !self.expects(vec![T![,], T![::]]) {
                    break;
                }
            }

            self.expect(T!["}"]);
        } else if !self.at(T![;]) {
            self.start_node(IMPORT_SEGMENT);
            self.ident();
        }

        self.finish_node();
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_empty_import,"import foo;"}
    test_parser! {parse_import_with_multiple_segments,"import foo::bar::baz;"}
    test_parser! {parse_import_with_brace,"import foo::{bar,baz};"}
    test_parser! {parse_import_with_brace_nested_segments,"import foo::{bar::baz,fubar};"}
}
