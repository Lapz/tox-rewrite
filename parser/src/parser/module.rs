use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_module(&mut self) {
        self.start_node(MOD_DEF);

        self.expect(T![mod]);

        self.ident();

        self.expect(T![;]);

        self.finish_node();
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_single_mod,"mod foo;"}
    test_parser! {parse_many_mods,"mod bar; mod foo;"}
}
