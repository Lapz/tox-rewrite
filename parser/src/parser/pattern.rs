use crate::parser::{Parser, Precedence, Restrictions};
use syntax::T;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_pattern(&mut self, allow_literal: bool) {
        match self.current() {
            T!["("] => self.parse_tuple_pattern(allow_literal),
            IDENT => self.parse_binding_pattern(),
            T![self] => self.bump(),
            T![_] => self.parse_placeholder_pattern(),
            e => {
                if allow_literal {
                    self.start_node(LITERAL_PAT);
                    self.parse_expression(Precedence::Primary, Restrictions::default());
                    self.finish_node()
                } else {
                    self.error(
                        "Expected a literal pattern",
                        format!(
                            "Expected a literal pattern instead found `{}` {:?}",
                            self.current_string(),
                            e
                        ),
                    )
                }
            }
        }
    }

    fn parse_tuple_pattern(&mut self, allow_literal: bool) {
        self.start_node(TUPLE_PAT);

        self.bump(); // Eat the `(`

        while !self.at(EOF) && !self.at(T![")"]) {
            self.parse_pattern(allow_literal);
            if !self.at(T![")"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![")"]);
        self.finish_node();
    }

    fn parse_placeholder_pattern(&mut self) {
        self.start_node(PLACEHOLDER_PAT);
        self.expect(T![_]);
        self.finish_node();
    }

    fn parse_binding_pattern(&mut self) {
        self.start_node(BIND_PAT);
        self.ident();
        self.finish_node();
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_placeholder_pattern,"fn main(_:i32) {}"}

    test_parser! {parse_tuple_pattern,"fn main((x,y):i32) {}"}

    test_parser! {parse_binding_pattern,"fn main(x:i32) {}"}

    test_parser! {parse_nested_tuple_pattern,"fn main((x,(y,_)):i32) {}"}
}
