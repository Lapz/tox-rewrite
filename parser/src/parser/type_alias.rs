use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_type_alias(&mut self, checkpoint: rowan::Checkpoint) {
        self.start_node_at(checkpoint, TYPE_ALIAS_DEF);

        self.expect(T![type]);

        self.ident();

        if self.at(L_ANGLE) {
            self.parse_type_params(false);
        }

        self.expect(EQ);

        self.parse_type();

        self.expect(SEMI);

        self.finish_node()
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_type_alias,"type Foo = i32;"}
    test_parser! {parse_type_alias_params,"type ParseResult<T> = Result<T,void>;"}
}
