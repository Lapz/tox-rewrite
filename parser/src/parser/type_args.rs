use crate::parser::Parser;
use crate::SyntaxKind::*;
use syntax::T;

impl<'a> Parser<'a> {
    pub(crate) fn parse_type_args(&mut self) {
        self.start_node(TYPE_ARG_LIST);

        self.bump();

        while !self.at(EOF) && !self.at(T![>]) {
            self.parse_type();

            if !self.at(T![>]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![>]);

        self.finish_node();
    }
}
