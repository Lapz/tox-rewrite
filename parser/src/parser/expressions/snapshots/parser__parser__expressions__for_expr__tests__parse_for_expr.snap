---
source: parser/src/parser/expressions/for_expr.rs
expression: "crate::utils::dump_debug(&parser_output)"
---
SOURCE_FILE@[0; 165)
  WHITESPACE@[0; 21) "\n                    "
  FN_DEF@[21; 165)
    FN_KW@[21; 23) "fn"
    WHITESPACE@[23; 24) " "
    NAME@[24; 28)
      IDENT@[24; 28) "main"
    PARAM_LIST@[28; 30)
      L_PAREN@[28; 29) "("
      R_PAREN@[29; 30) ")"
    WHITESPACE@[30; 31) " "
    BLOCK_EXPR@[31; 165)
      BLOCK@[31; 165)
        L_CURLY@[31; 32) "{"
        WHITESPACE@[32; 57) "\n                     ..."
        EXPR_STMT@[57; 163)
          FOR_EXPR@[57; 163)
            FOR_KW@[57; 60) "for"
            L_PAREN@[60; 61) "("
            LET_STMT@[61; 71)
              LET_KW@[61; 64) "let"
              WHITESPACE@[64; 65) " "
              BIND_PAT@[65; 66)
                NAME@[65; 66)
                  IDENT@[65; 66) "x"
              WHITESPACE@[66; 67) " "
              EQ@[67; 68) "="
              WHITESPACE@[68; 69) " "
              LITERAL@[69; 71)
                INT_NUMBER@[69; 71) "10"
            SEMI@[71; 72) ";"
            WHITESPACE@[72; 73) " "
            BIN_EXPR@[73; 78)
              IDENT_EXPR@[73; 74)
                NAME@[73; 74)
                  IDENT@[73; 74) "x"
              L_ANGLE@[74; 75) "<"
              WHITESPACE@[75; 76) " "
              LITERAL@[76; 78)
                INT_NUMBER@[76; 78) "10"
            SEMI@[78; 79) ";"
            WHITESPACE@[79; 80) " "
            BIN_EXPR@[80; 85)
              IDENT_EXPR@[80; 81)
                NAME@[80; 81)
                  IDENT@[80; 81) "x"
              EQ@[81; 82) "="
              BIN_EXPR@[82; 85)
                IDENT_EXPR@[82; 83)
                  NAME@[82; 83)
                    IDENT@[82; 83) "x"
                PLUS@[83; 84) "+"
                LITERAL@[84; 85)
                  INT_NUMBER@[84; 85) "1"
            R_PAREN@[85; 86) ")"
            WHITESPACE@[86; 87) " "
            BLOCK_EXPR@[87; 163)
              BLOCK@[87; 163)
                L_CURLY@[87; 88) "{"
                WHITESPACE@[88; 117) "\n                     ..."
                EXPR_STMT@[117; 162)
                  CALL_EXPR@[117; 137)
                    IDENT_EXPR@[117; 124)
                      NAME@[117; 124)
                        IDENT@[117; 124) "println"
                    ARG_LIST@[124; 137)
                      L_PAREN@[124; 125) "("
                      LITERAL@[125; 136)
                        STRING@[125; 136) "\"it works \""
                      R_PAREN@[136; 137) ")"
                  WHITESPACE@[137; 162) "\n                     ..."
                R_CURLY@[162; 163) "}"
        SEMI@[163; 164) ";"
        R_CURLY@[164; 165) "}"

